src/socket.rs

```diff
@@ -110,6 +75,54 @@ pub struct Socket {
 }
 
 impl Socket {
+    /// Creates a new socket ready to be configured.
+    ///
+    /// This function corresponds to `socket(2)` on Unix and `WSASocketW` on Windows.
+    ///
+    /// # Notes
+    ///
+    /// On Unix-like systems, the close-on-exec flag is set on the new socket.
+    /// Additionally, on macOS X and iOS the `SOCK_NOSIGPIPE` option is set.
+    /// On Windows, the socket handle is made non-inheritable.
+    ///
+    /// If possible, the close-on-exec flag will be set atomically when the socket is created.
+    /// However, on some platforms the flag is to be set with a separate syscall,
+    /// which leaves a small window for race conditions.
+    ///
+    /// Use [`Self::new_raw`] if you don't want these flags and options to be set.
+    pub fn new(domain: Domain, ty: Type, protocol: Option<Protocol>) -> io::Result<Socket> {
+        #[cfg(any(
+            target_os = "android",
+            target_os = "dragonfly",
+            target_os = "freebsd",
+            target_os = "linux",
+            target_os = "netbsd",
+            target_os = "openbsd",
+        ))]
+        {
+            Self::new_raw(domain, ty._cloexec(), protocol)
+        }
+
+        #[cfg(windows)]
+        {
+            Self::new_raw(domain, ty._no_inherit(), protocol)
+        }
+
+        #[cfg(not(any(
+            target_os = "android",
+            target_os = "dragonfly",
+            target_os = "freebsd",
+            target_os = "linux",
+            target_os = "netbsd",
+            target_os = "openbsd",
+            target_os = "windows",
+        )))]
+        {
+            // Fix portability flags after creating the socket.
+            fixup_socket(Self::new_raw(domain, ty, protocol)?)
```

Comment:
`fixup` is not a good name as it indicates the socket is somehow broken.

Reply:
I'd argue it is without those flags, but sure, will rename.